// @primuslabs/zktls-js-sdk/dist/classes/AttRequest.js
import { getInstanceProperties } from '../utils';
import { v4 as uuidv4 } from 'uuid';
export class AttRequest {
    appId;
    attTemplateID;
    userAddress;
    timestamp;
    attMode;
    attConditions;
    additionParams;
    requestid;
    backUrl;
    computeMode;
    noProxy;
    allJsonResponseFlag;
    /** Attestation polling timeout in milliseconds. */
    timeout;
    constructor(baseAttestationParams) {
        const { appId, attTemplateID, userAddress, timeout } = baseAttestationParams;
        this.appId = appId;
        this.attTemplateID = attTemplateID;
        this.userAddress = userAddress;
        this.timestamp = +new Date();
        this.attMode = {
            algorithmType: 'proxytls',
            resultType: 'plain'
        }; // TODO
        this.requestid = uuidv4();
        this.backUrl = "";
        this.computeMode = "normal";
        this.noProxy = true;
        this.allJsonResponseFlag = 'false';
        if (timeout !== undefined) {
            this.timeout = timeout;
        }
    }
    setAdditionParams(additionParams) {
        this.additionParams = additionParams;
    }
    setAttMode({ algorithmType, resultType = 'plain' }) {
        this.attMode = {
            algorithmType,
            resultType
        };
    }
    setAttConditions(attConditions) {
        this.attConditions = attConditions;
    }
    setBackUrl(url) {
        this.backUrl = url;
    }
    setComputeMode(computeMode) {
        this.computeMode = computeMode;
    }
    setNoProxy(noProxy) {
        this.noProxy = noProxy;
    }
    setAllJsonResponseFlag(allJsonResponseFlag) {
        this.allJsonResponseFlag = allJsonResponseFlag === 'true' ? 'true' : 'false';
    }
    toJsonString() {
        return JSON.stringify(getInstanceProperties(this));
    }
}
